### PR TITLE
Fix multiple bugs in matching ##signatures

### PR DESCRIPTION
**Pull request by swoops**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [x] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

This refactors out multiple bugs in the matching of signatures. When a signature is matched, either `fcnMatchCB` or `searchHitCB` is called. This previous lead to `addFlag` where there were a few bugs. These bugs prevented the function from being renamed, having types applied and resulted in some leaks.

There is still a bug where types are not applied because the `RSignItem` passed to `searchHitCB` does not have a `types` member. This is due to a bug in `r_sign_item_dup` and is addressed in #17608.

**Test plan**
```
$ r2 static 
 -- Did you know that radare2 can decompile to assembly?
[0x00401c60]> s 0x0041dec0
[0x0041dec0]> af
[0x0041dec0]> afs
void fcn.0041dec0 (int64_t arg1);
[0x0041dec0]> zo ./out.sdb 
[0x0041dec0]> z*
zs *
za sym.__strdup b f30f1efa4154554889fd4883ec08e83d69f9ff4c8d60014c89e7e8d167f9ff4885c074154883c4084c89e24889ee4889c75d415ce98769f9ff5a31c05d415cc3:ffffffffffffffffffffffffffffff00000000ffffffffffffffff00000000ffffffff00ffffffffffffffffffffffffffffffffff00000000ffffffffffffff
za sym.__strdup g cc=2 nbbs=3 edges=3 ebbs=1 bbsum=64
za sym.__strdup o 0x0008fb30
za sym.__strdup v r110
za sym.__strdup t func.sym.__strdup.args=1 func.sym.__strdup.arg.0="uint32_t,AAAA"
za sym.__strdup h d8b82299783a827712112871062c12f183da2579815bbb644bec13b71fac5618
[0x0041dec0]> z.
[+] searching 0x0041dec0 - 0x0041dfc0
[+] searching function metrics
hits: 1
[0x0041dec0]> afs
void sym.__strdup (uint32_t AAAA);
[0x0041dec0]> 
```
Note: I use `z.` here b/c it is fast and easy to show. This PR alone won't make `z.` assign types correctly because of the other bug mentioned above. 
**Closing issues**

...

### CODE REVIEW

File: libr/core/cmd_zign.c [hunk 1]
```diff
@@ -675,112 +675,69 @@ struct ctxSearchCB {
...
-			free (fcnstr_copy);
+		if (!(alltypes = r_str_appendf (alltypes, "%s\n", type))) {
+			free (alltypes);
+			free (start);
+			return false;
 		}
-		free (fcnstr);
 	}
-	name = r_name_filter2 (r_str_newf ("%s.%s.%s_%d", zign_prefix, prefix, it->name, count));
+	r_str_remove_char (alltypes, "\"");
```
Comment (ret2libc) [inline libr/core/cmd_zign.c:725]:
```suggestion
	r_str_remove_char (alltypes, '"');
```

File: libr/core/cmd_zign.c [hunk 2]
```diff
@@ -675,112 +675,69 @@ struct ctxSearchCB {
...
+
+static void apply_flag(RCore *core, RSignItem *it, ut64 addr, int size, int count, const char *prefix, bool rad) {
+	const char *zign_prefix = r_config_get (core->config, "zign.prefix");
+	char *name = r_str_newf ("%s.%s.%s_%d", zign_prefix, prefix, it->name, count);
 	if (name) {
 		if (rad) {
-			r_cons_printf ("f %s %d @ 0x%08"PFMT64x"\n", name, size, addr);
-			if (it->realname) {
-				r_cons_printf ("\"afn %s @ 0x%08"PFMT64x"\"\n", it->realname, addr); // XXX command injection
+			char tmp = r_name_filter2 (name);
```
Comment (ret2libc) [inline libr/core/cmd_zign.c:737]:
```suggestion
			char *tmp = r_name_filter2 (name);
```

File: libr/core/cmd_zign.c [hunk 3]
```diff
@@ -675,112 +675,69 @@ struct ctxSearchCB {
...
-					if (arg[j] == ',') {
-						arg[j] = ' ';
-					}
-				}
-				r_list_append (args, r_str_ndup (arg + 1,
-					strlen (arg) - 2));
-				free (arg);
-			}
-			i++;
+static bool apply_name(RCore *core, RAnalFunction *fcn, RSignItem *it, bool rad) {
```
Comment (ret2libc) [inline libr/core/cmd_zign.c:678]: the return value is never used and it's always true. just return void.

File: libr/core/cmd_zign.c [hunk 4]
```diff
@@ -675,112 +675,69 @@ struct ctxSearchCB {
...
-			r_str_trim_tail (arg);
-			const char *fcn_name = strrchr (arg, ' ');
-			// __setFunctionName() ; cmd_anal.c:2535 ; Expand into R_API function
-			free (fcn->name);
-			fcn->name = strdup (fcn_name + 1);
-			if (core->anal->cb.on_fcn_rename) {
-				core->anal->cb.on_fcn_rename (core->anal, core->anal->user, fcn, fcn->name);
-			}
-			r_anal_str_to_fcn (core->anal, fcn, fcnstr);
+static bool apply_types(RCore *core, RAnalFunction *fcn,  RSignItem *it) {
```
Comment (ret2libc) [inline libr/core/cmd_zign.c:701]: also here

**Comment by ret2libc** [comment]
How can I reproduce the test case? Could you provide this "static" binary at least here.

**Comment by swoops** [comment]
Using `ls` from the test repo

```r2 bins/elf/ls
 -- I love gradients.
 [0x00005ae0]> s 0x00012580
[0x00012580]> af
[0x00012580]> afs
void fcn.00012580 (int64_t arg1);
[0x00012580]> zo /tmp/test.sdb 
[0x00012580]> z.
[+] searching 0x00012580 - 0x00012680
[+] searching function metrics
hits: 2
[0x00012580]> afs
void sym.renamed_value (int64_t AAAA);
[0x00012580]> z* ~ t 
za sym.renamed_value t func.sym.renamed_value.args=1 func.sym.renamed_value.arg.0="int64_t,AAAA"
```

I am attaching the .sdb file as a zip. Sadly you can't add the signature with za because of another bug I am working on fixing. Once that is fixed I should be able to add very simple regression tests for this.

**Review by ret2libc — APPROVED** [review]
Thanks!

File: libr/core/cmd_zign.c [hunk 5]
```diff
@@ -675,7 +675,7 @@ struct ctxSearchCB {
 	const char *prefix;
 };
 
-static bool apply_name(RCore *core, RAnalFunction *fcn, RSignItem *it, bool rad) {
+static void apply_name(RCore *core, RAnalFunction *fcn, RSignItem *it, bool rad) {
 	r_return_val_if_fail (core && fcn && it && it->name, false);
```
Comment (trufae) [inline libr/core/cmd_zign.c]: how to return a val if its void

File: libr/core/cmd_zign.c [hunk 6]
```diff
@@ -684,24 +684,23 @@ static bool apply_name(RCore *core, RAnalFunction *fcn, RSignItem *it, bool rad)
...
 	r_anal_function_rename (fcn, name);
 	if (core->anal->cb.on_fcn_rename) {
 		core->anal->cb.on_fcn_rename (core->anal, core->anal->user, fcn, name);
 	}
-	return true;
 }
 
-static bool apply_types(RCore *core, RAnalFunction *fcn,  RSignItem *it) {
+static void apply_types(RCore *core, RAnalFunction *fcn, RSignItem *it) {
 	r_return_val_if_fail (core && fcn && it && it->name, false);
```
Comment (trufae) [inline libr/core/cmd_zign.c]: same here

**Review by trufae — CHANGES_REQUESTED** [review]
imho it will better to check for errors in the callers than returning void

File: libr/core/cmd_zign.c [hunk 7]
```diff
@@ -359,20 +359,20 @@ static bool addVarsZign(RCore *core, const char *name, const char *args0, int na
...
-       return retval;
+	int i = 0;
+	if (nargs < 1) {
+		eprintf ("error: invalid syntax\n");
+		return false;
+	}
+
+	RList *types = r_list_newf ((RListFree)free);
+	for (i = 0; i < nargs; i++) {
+		r_list_append (types, r_str_new (r_str_word_get0 (args0, i)));
```
[the remark this answers is not in the document]
Comment (swoops) [reply libr/core/cmd_zign.c]: I will use strdup. There is some other problems with that code which may require it to be changed more substantially.

**Review by trufae — APPROVED** [review]

**Comment by trufae** [comment]
Damn, I think it will be good to have some tests for those fixes. Can you make a separte PR for them? Thanks

**Pull request merged**